ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/DeduplicationPeriod.scala

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.v2
+
+import java.time.{Duration, Instant}
+
+import com.daml.ledger.participant.state.v1.Offset
+
+/** Specifies the deduplication period for a command submission.
+  *
+  * @see com.daml.ledger.participant.state.v2.ReadService.stateUpdates for the deduplication guarantee
+  */
+sealed trait DeduplicationPeriod extends Product with Serializable
+
+object DeduplicationPeriod {
+
+  /** The `offset` defines the start of the deduplication period. */
+  case class DeduplicationOffset(offset: Offset) extends DeduplicationPeriod
+
+  /** The instant `since` specifies the point in time where deduplication starts.
+    * This point in time is measured on some unspecified clock on the participant or the Daml ledger.
+    */
+  case class DeduplicationTimepoint(since: Instant) extends DeduplicationPeriod
+
+  /** The length of the deduplication window, which ends when the [[WriteService]] or underlying Daml ledger processes
+    * the command submission. The window is measured on some unspecified clock on the participant or the Daml ledger.
+    *
+    * @throws java.lang.IllegalArgumentException if the `duration` is negative
+    */
+  case class DeduplicationWindow(duration: Duration) extends DeduplicationPeriod {
+    require(!duration.isNegative, s"The deduplication window must not be negative: $duration")
```

Comment:
@andreaslochbihler-da : this exception would likely be raised by the request parsing code in the API server. If we do this, will it be translated to the righ gRPC status code?

I have a hunch that the control flow would be easier to understand if parsing was total and the validation of the `DeduplicationPeriod` checked this constraint.

Reply:
With defensive programming, the parsing code should responsible for checking this precondition before invoking the constructor. We can add a suitable parsing method that does that for the companion object.  (That's how we're doing it in Canton.) The benefit of having the `require` here is that all users of the class can rely on the duration being non-negative.

Reply:
Thank you for the explanation. That makes sense.